src/type_checker.cpp

```diff
@@ -140,6 +140,51 @@ void TypeChecker::Visit(ContinueStmtNode& continue_stmt) {
   }
 }
 
+namespace {
+/// @brief A shared state to convey the presence of a default label in a switch
+/// statement.
+/// @note To allow nested switch statements, the state is stacked.
+auto
+    switch_already_has_default  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+    = std::vector<bool>{};
+}  // namespace
+
+void TypeChecker::Visit(SwitchStmtNode& switch_stmt) {
+  switch_stmt.ctrl->Accept(*this);
+  if (switch_stmt.ctrl->type != ExprType::kInt) {
+    // TODO: statement requires expression of integer type
+  }
+  body_types.push_back(BodyType::kSwitch);
+  switch_already_has_default.push_back(false);
+  switch_stmt.stmt->Accept(*this);
+  switch_already_has_default.pop_back();
+  body_types.pop_back();
+  // TODO: No two of the case constant expressions in the same switch statement
+  // shall have the same value (we need constant expression support on this).
+}
+
+void TypeChecker::Visit(CaseStmtNode& case_stmt) {
+  if (!IsInBodyOf(BodyType::kSwitch)) {
+    // TODO: 'case' statement not in switch statement
+  }
+  case_stmt.expr->Accept(*this);
+  if (case_stmt.expr->type != ExprType::kInt) {
+    // TODO: expression is not an integer constant expression
+  }
+  case_stmt.stmt->Accept(*this);
+}
+
+void TypeChecker::Visit(DefaultStmtNode& default_stmt) {
+  if (!IsInBodyOf(BodyType::kSwitch)) {
+    // TODO: 'default' statement not in switch statement
+  }
+  assert(!switch_already_has_default.empty());
+  if (switch_already_has_default.back()) {
```

Comment:
Wouldn't `default_stmt` need to push a new `true` node into `switch_already_has_default`?

Reply:
Sorry for the confusion. The current state should be set to `true` instead of pushing a new one.

Reply:
Also our testing infrastructure doesn't support testing on failed cases. :sob:
I'll skip adding the test case.